hier.7: deep rewrite for 14

concussious opened the pull request:
Rewrite hier in the system manual, because it is the entry point to the system manual mentioned in the default motd. As the handbook has been moved to ports, this is in band in base and always suggested to users. This is the first impression of the installed system to the attentive new user.

- add spdx identifier
- make document description more descriptive to improve apropos and reader comprehension
- zap synopsis, this was not a synopsis in the sense of the term used in the rest of the manual
- sentences for descriptions, I think this is more consistent with the rest of the writing in the system manual
- attempt to improve some descriptions while keeping them as concise as I could conceive, thinking of new users using the / key to search
- add many entries with the base principle: most concise explanation for indexing, as a reflection of it's position as the instruction on MOTD
- especially mention /boot/fonts. This is a really cool feature other operating systems do not have, and it's only 1 directory deep.
- especially completely (to my knowlege) mention default logs, except /var/log/debug, I think this has major effect on new users speed at comprehending the system
- Remove (to my knowledge) nonexistent parts of the system or moved parts of the system, such as:
- Add a little to description of /usr/local/. This is a really key concept of FreeBSD for users, and since the handbook was moved to here, that must continue being mentioned. The system manual and the handbook are the primary, official documentation.
- mention /var/log/Xorg.0.log. Yes, it's in ports, but significant parts of the system are built around this, and anyone who deliberately doesn't have this already knows what this is.
- Added year to § history. This is a small change that adds great value as it's actually profound for younger people who don't automatically know when unix v7 era was. The consistency of tradition attracted me to FreeBSD and simply is one of the uncontested advantages of this platform.

There are ~~6~~ 5 listed with no explanation, I failed to identify what they do. That said, there always have been and the proposed change leaves those entries at a smaller ratio than there was before. Hopefully someone will weigh in and it will be the most complete it's been this century for the release. Those are:

- ~~/entropy~~   # Resolved, thanks @stesser!
- /dev/drm   # I can't find this anywhere
- /var/log/debug.log 
- /usr/share/misc/fonts   # who is making this? To remove it's listing after an entire lifetime, it is heritage tier, nobody knows?
- /var/spool/lock
- /var/spool/opielocks

Thank you all for your time and teaching.
@mhorne asked me to tag if I submit such a thing

stesser commented on share/man/man7/hier.7:
```diff
@@ -25,432 +28,747 @@
+USB busses.
 .El
+.It Pa /entropy
+Stored entropy file; see _____.
```
The entropy file is used to provide initial state to the random number generator.
It is created and used by /etc/rc.d/random.
Depending on settings in /etc/rc.conf it may also be moved into some other directory or can be provided in a special place for the first boot (e.g. to make sure individual SSH host keys are generated on a freshly installed system or VM).
See random_start() in /etc/rc.d/random for details ...

concussious replied:
Awesome, thank you so much. Is there a knob for that we can mention in rc.conf(5)?
I feel like it'd be consistent to put the entropy file in /var.

stesser replied:
> Is there a knob for that we can mention in rc.conf(5)? I feel like it'd be consistent to put the entropy file in /var.

Providing entropy to the random generator supports many scenarios and therefore offers quite a number of options that can be configured in rc.conf:
```
entropy_boot_file="/boot/entropy"	# Set to NO to disable very early
			# (used at early boot time) entropy caching through reboots.
entropy_file="/entropy"	# Set to NO to disable late (used when going multi-user)
			# entropy through reboots.
			# /var/db/entropy-file is preferred if / is not avail.
entropy_dir="/var/db/entropy" # Set to NO to disable caching entropy via cron.
```
The comments explain the reasons for 3 different locations:

- _entropy_boot_file_ may be on a separate boot file system (not necessarily on the same partition as the root file system), it could for example also be provided by a network boot server
- _entropy_file_ is on the root file system and thus available before any user processes that need random numbers have been started (e.g. the initial start of sshd will need entropy to create an unpredictable host key)
- _entropy_dir_ contains additional entropy files with random data saved a few times per hour while the system is up and running (but /var may be on a file system different from the root file system and thus possibly not available at the start of going multi-user)

The contents of the _entropy_file_ and _entropy_boot_file_ will be overwritten with random data when going multi-user. An additional entropy file named /var/db/entropy-file is hard-coded in the /etc/rc.d/random. It will be read in if present (in addition to _entropy_file_), but will not be overwritten with new random data.

If modified, the value assigned to _entropy_boot_file_ in rc.conf needs to be added to loader.conf, too, since that file is read in by the loader and then passed to the kernel (see the definitions and comments in /boot/defaults/loader.conf):
```
# See rc.conf(5). The entropy_boot_file config variable must agree with the
# settings below.
entropy_cache_load="YES"		# Set this to NO to disable loading
					# cached entropy at boot time
entropy_cache_name="/boot/entropy"	# Set this to the name of the file
```

grahamperrin commented:
Context: FreeBSD bug [261349 – Modernise hier(7)](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=261349)

grahamperrin commented:
The editions that led to what's currently <https://man.freebsd.org/cgi/man.cgi?query=hier&sektion=7&manpath=FreeBSD+15.0-CURRENT> intentionally had lowercase beginnings to phrases that are adjacent to paths, with less need to punctuate with full stops etc..

Maybe not consistent with some other pages, however the [DESCRIPTION](https://man.freebsd.org/cgi/man.cgi?query=hier&sektion=7&manpath=FreeBSD+15.0-CURRENT#DESCRIPTION) section of hier(7) is entirely tabular, which is probably unusual for a DESCRIPTION.

----

For me, the addition of uppercase and punctuation makes parts of the page considerably more difficult to digest, especially in windows that are narrow.

grahamperrin reviewed (changes requested):
The two typos were found at a glance, I have not checked the rest of the page for similar errors.

concussious commented on share/man/man7/hier.7:
```diff
@@ -462,161 +826,277 @@ The layout of the source tree is described by the top-level
+see
+.Xr msgs 1 .
 .It Pa preserve/
-unused, present for historical reasons
+Temporary home of files preserved after an accidental death of the
+.Xr ex 1
+text editor.
```
Thanks, I didn't see those.

mhorne commented on share/man/man7/hier.7:
```diff
@@ -462,161 +826,277 @@ The layout of the source tree is described by the top-level
 .Pp
-.Bl -tag -width "utx.lastlogin" -compact
+.Bl -tag -width 14n -compact
+.It Pa Xorg.0.log
```
I am not sure what to think of this yet...

This page documents directories, not files. Yet I do understand the utility of adding it here...

concussious replied:
The convention is to document directories and not files, but afaik that isn't an explicit rule. I interpreted the term hierarchy as neutrally inclusive to either. I think the ratio of utility to maintenence burden of filling out /var/log is actually quite impressive.

mhorne reviewed (changes requested):
Hi,

I must apologize for my long silence on this review. I took a long vacation through September and have been working on other things since I got back.

At the high level, the diff presented here is far too big, making it difficult to review. In order for this to be acceptable for merge, I must request that it be split into at least three separate commits by category (in order):
 1. Style/formatting changes
 2. Changes/removal to existing content
 3. Addition of new content

I agree with @grahamperrin that the formatting choice to capitalize each item description and give it a full-stop is undesirable. Doing so treats each description like a sentence, and many of them are not complete sentences.

Otherwise, see my inline comments.

Thanks for your work so far! I appreciate that you have spent a lot of time on this already.

bsdimp commented:
Modulo the dates in the .Dd lines and my removal of the SCCS tags, this rebases forward.
It looks like the conversation here was favorable.
Is there some reason it hasn't been committed?

bsdimp commented:
This needs a real email, not a noreply github one for the commits (or some of them)

concussious commented:
> Modulo the dates in the .Dd lines and my removal of the SCCS tags, this rebases forward. It looks like the conversation here was favorable. Is there some reason it hasn't been committed?

* I (accidentally) added a lot of content that I didn't see was previously removed, which seems to be a blocker. I don't know what to do about it yet. I added it because I think it is helpful, and I think the contents of these directories changes very little over what seems like massive amounts of time. No agreement has been reached yet. 

* I made the descriptions into sentences. I agree that this needs to be reverted, sorry for the delay.

* I changed the formatting so I could see what I'm doing in the VT console/tmux/nvi+mantra workflow. This has been requested to be a separate commit, but I don't think that's reasonable. If the new content is added with the same formatting the line wrapping would be broken. No agreement has been reached yet.

I'm sorry for the delay, internet access is not a regular part of my life.

igalic commented:
> This needs a real email, not a noreply github one for the commits (or some of them)

https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/setting-your-commit-email-address

you need to set the commit email in your git cli, then go back rewrite the author: https://stackoverflow.com/questions/750172/how-do-i-change-the-author-and-committer-name-email-for-multiple-commits#1320317

concussious commented:
> > This needs a real email, not a noreply github one for the commits (or some of them)

Did that work? If it did, this is still not ready for merging yet. I think hier.7 is really important and I'd like this to be more perfect before it's accepted.

concussious commented on share/man/man7/hier.7:
```diff
@@ -25,432 +28,793 @@
+Virtual memory;
+see
+.Xr vmstat 8 .
+.El
```
I'm sorry, I actually failed to see that. Removed in latest commit.

concussious commented on share/man/man7/hier.7:
```diff
@@ -25,432 +28,793 @@
+.Pp
+.It Pa dtrace/
+Scripts for the Dynamic Tracing Compiler, see
+.Xr dtrace 1 .
```
My mistake, I agree with your reasoning, sorry for the noise.

concussious commented:
I'm so sorry for the noise, I renamed the branch and didn't know it would break the PR.

bsdimp commented:
Pushed this in.... Split it up a little, though. I didn't provide good feedback early... but will in other pull commits
Merged.
